traffic_router/ultimate-test-harness/http_test.go

```diff
@@ -0,0 +1,497 @@
+				redirects += <-redirectsChannels[threadIndex]
+				failures += <-failuresChannels[threadIndex]
+			}
+			protocol := "IPv6"
+			if isIPv4 {
+				protocol = "IPv4"
+			}
+			var passed string
+			requestsPerSecond := redirects / benchmark.BenchmarkTime
+			if requestsPerSecond > benchmark.RequestsPerSecondThreshold {
+				passedTests++
+				passed = "Yes"
+			} else {
+				failedTests++
+				passed = "No"
+			}
+			fmt.Fprintf(writer, "%s\t%s\t%s\t%s\t%d\t%d\t%d\n", trafficRouter.Hostname, protocol, trafficRouter.DSHost, passed, requestsPerSecond, redirects, failures)
+			writer.Flush()
+		}
+	}
+	summary := fmt.Sprintf("%d out of %d load tests passed", passedTests, passedTests+failedTests)
+	if failedTests < 1 {
+		t.Logf(summary)
+	} else {
+		t.Fatal(summary)
+	}
+}
+
+func (b *Benchmark) Run(t *testing.T, redirectsChannel chan int, failuresChannel chan int, trafficRouterIndex int, trafficRouterURL string, ipAddressIndex int) {
+	paths := generatePaths(b.PathCount, b.MaxPathLength)
+	stopTime := time.Now().Add(time.Duration(b.BenchmarkTime) * time.Second)
+	redirects, failures := 0, 0
+	var req *http.Request
+	var resp *http.Response
+	var err error
+	httpClient := http.Client{
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+	trafficRouter := b.TrafficRouters[trafficRouterIndex]
+	for time.Now().Before(stopTime) {
+		requestURL := trafficRouterURL + paths[rand.Intn(len(paths))]
+		if req, err = http.NewRequest("GET", requestURL, nil); err != nil {
+			t.Fatalf("creating GET request to Traffic Router '%s' (IP address %s): %s",
+				trafficRouter.Hostname, trafficRouter.IPAddresses[ipAddressIndex], err.Error())
+		}
+		req.Header.Set("User-Agent", UserAgent)
+		if trafficRouter.ClientIP != "" {
+			req.Header.Set(tc.X_MM_CLIENT_IP, trafficRouter.ClientIP)
+		}
+		req.Host = trafficRouter.DSHost
+		resp, err = httpClient.Do(req)
+		if err == nil && resp.StatusCode >= http.StatusMultipleChoices && resp.StatusCode < http.StatusBadRequest {
+			redirects++
+		} else {
+			failures++
+		}
+	}
+	redirectsChannel <- redirects
+	failuresChannel <- failures
+}
+
+func generatePaths(pathCount, maxPathLength int) []string {
+	const alphabetSize = 26 + 26 + 10
+	alphabet := make([]rune, alphabetSize)
+	index := 0
+	for char := 'A'; char <= 'Z'; char++ {
+		alphabet[index] = char
+		index++
+	}
+	for char := 'a'; char <= 'z'; char++ {
+		alphabet[index] = char
+		index++
+	}
+	for char := '0'; char <= '9'; char++ {
+		alphabet[index] = char
+		index++
+	}
+	paths := make([]string, pathCount)
+	for index = 0; index < pathCount; index++ {
+		pathLength := rand.Intn(maxPathLength)
+		generatedURL := make([]rune, pathLength)
+		for runeIndex := 0; runeIndex < pathLength; runeIndex++ {
+			generatedURL[runeIndex] = alphabet[rand.Intn(alphabetSize)]
+		}
+		paths[index] = string(generatedURL)
+	}
+	return paths
+}
+
+func BenchmarkHttpDSes(b *testing.B) {
```

Comment:
What's the purpose of this benchmark?

Reply:
Oops, that snuck in there. Removed in cac9a3599c.